Tooltip: improve perf by preventing all render until after delay's done.

aneeshack4 opened the pull request:
#### Pull request checklist

- [x] Addresses an existing issue: Fixes #9321
- [x] Include a change request file using `$ npm run change`

#### Description of changes

We're currently relying on CSS to delay the visibility of the tooltip. By wrapping the Callout inside Tooltip with <DelayedRender> and passing in the delay calculated in styles, we can make sure nothing renders/gets added to the DOM until after the delay's fully done. 

#### Focus areas to test

Play around with the Tooltip examples in the PR deploy link - make sure nothing gets added to the DOM until after each delay's done as you mouse over different tooltip areas.


###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/9341)

LiangMingChen commented:
> Thanks for Fixing,
> if we fix the by Fix render, then we don't need
> Line 45 getClassName(with delay argument), we should remove the animation
> https://github.com/OfficeDev/office-ui-fabric-react/blob/8a471edeec02db18f201db9790da374c573f9f2b/packages/office-ui-fabric-react/src/components/Tooltip/Tooltip.styles.ts#L16
> 
> ```
>     animationDelay: '300ms',
> ```

Tooltip's delay is  a enum (0, 1, 2)  I'd image the utility delayRender is taking a ms.
Probably want to change to 1 second by default.

Here is browser default 'title'  (even though it is various by default.)
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/title
(when you hover,  it feels like 500ms to 1ms delay to show the hover. -- there is no doc on how native is implemented.)

micahgodbolt commented on packages/office-ui-fabric-react/src/components/Tooltip/Tooltip.base.tsx:
```diff
@@ -47,24 +47,26 @@ export class TooltipBase extends React.Component<ITooltipProps, any> {
-          role="tooltip"
-          onMouseEnter={this.props.onMouseEnter}
-          onMouseLeave={this.props.onMouseLeave}
+      <DelayedRender delay={delay}>
```
delay value is an enum. you'll need to do a quick case switch for the 3 matches. 0, 300, 500ms

micahgodbolt replied:
Also (as mentioned) remove it from the styles file

micahgodbolt replied:
This needs to get addressed before your tests pass.

aneeshack4 replied:
I thought I addressed this with the following: https://github.com/OfficeDev/office-ui-fabric-react/pull/9341/files#diff-00719f3cbd47c1407b8d401275c4292eR48 and I did remove it from the styles file altogether: https://github.com/OfficeDev/office-ui-fabric-react/pull/9341/files#diff-797904568f02429664f4dc0012b330b9L16

aneeshack4 commented:
Addressed the feedback in the latest commit & pulled down the latest changes to resolve merge conflicts but I'm still getting these two weird errors when I do `npm run update-snapshots` I looked at the test file and I don't understand why the tooltip would be null in that case. @micahgodbolt do you know what's going wrong?

LiangMingChen commented:
> I looked at the test file and I don't understand why the tooltip would be null in that case. @micahgodbolt do you know what's going wrong?

That's probably becuase now you delay render using the helper function.  You might want to look into how the delayRender HOC's test to wait for that delay to have &lt;CallOut&gt; rendered.
You should only able to find the &lt;CallOut&gt; only after it render.  
When it failed to fine &lt;CallOUt&gt;, it will have null return,
thus callout.prop('') will crash as call out is null

leddie24 commented:
Hi @aneeshack4 , just wanted to ping on this PR, it looks like it's been a few days since it's seen activity.

aneeshack4 commented:
@leddie24 Ah yes last time I checked there were no merge conflicts so I was just waiting for approval to merge. I'll resolve these.

KevinTCoughlin commented:
Hmmm.. one concern here is that we no longer have snapshot coverage of the `Tooltip` component, see https://github.com/OfficeDev/office-ui-fabric-react/pull/9341/files#diff-ec79dcef7c696917b8ad03c3d69d1ac6. Is there a way we can still have snapshot coverage?

micahgodbolt commented:
oh snapshot could probably be return by specifying delay of zero. We weren't having much luck testing things that used DelayedRender. Probably some async updates that need to happen to test those with delay

KevinTCoughlin commented:
> oh snapshot could probably be return by specifying delay of zero. We weren't having much luck testing things that used DelayedRender. Probably some async updates that need to happen to test those with delay

Cool, I suggest we do that if possible so we retain coverage :).

natalieethell commented:
I'm seeing a run-time error when trying to hover over any of the tooltips on the [example page](http://fabricweb.z5.web.core.windows.net/pr-deploy-site/refs/pull/9341/merge/fabric-website-resources/dist/index.html#/examples/Tooltip). It might be something related to passing `{RenderedCallout}` as a child of `DelayedRender`? Here's the error: https://reactjs.org/docs/error-decoder.html/?invariant=143
